### DISCUSSION TITLE
`DataAttribute` with disposable data

### DISCUSSION THREAD

**Discussion by rick-palmsens** [xunit/xunit#2610] [Question] [answered]
I am writing some integration tests where we have multiple hardware devices connected to the build server, and we run the test for each device separately. I did the whole `SupportsDiscoveryEnumeration` thingy to prevent access to the devices during discovery.

But now I'm dealing with the following issue: how do I dispose my data? At first I did basically this:
```
yield return device;
device.Dispose();
```

But then it turned out that `GetData()` is completely enumerated BEFORE running any tests, so my device ended up being disposed before the test.

So at this point I don't know where I can dispose of my connection. Especially with physical hardware it is important to properly close the connection. Is there any kind of extension point where I can release data generated by a `DataAttribute`?

**Comment by bradwilson** [comment, marked as answer]
The fact that your data items implement `IDisposable` is enough: we will automatically call `Dispose()` when we're done with the data item. There's no need for you to try to call `Dispose()` yourself.
